pkg/hub/controllers/finalizercontroller/add_finalizer_controller.go

```diff
@@ -0,0 +1,171 @@
+package finalizercontroller
+
+import (
+	"context"
+	"strings"
+
+	clusterv1listers "github.com/open-cluster-management/api/client/cluster/listers/cluster/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	rbacv1informers "k8s.io/client-go/informers/rbac/v1"
+	rbacv1client "k8s.io/client-go/kubernetes/typed/rbac/v1"
+	rbacv1listers "k8s.io/client-go/listers/rbac/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+const (
+	manifestWorkFinalizer = "cluster.open-cluster-management.io/manifest-work-cleanup"
+	keySuffix             = ":spoke-work"
+)
+
+// AddFinalizerController is to add the cluster.open-cluster-management.io/manifest-work-cleanup
+// finalizer on role/rolebinding for work agent in spoke cluster namespace.
+type AddFinalizerController struct {
+	roleLister        rbacv1listers.RoleLister
+	roleBindingLister rbacv1listers.RoleBindingLister
+	rbacClient        rbacv1client.RbacV1Interface
+	clusterLister     clusterv1listers.SpokeClusterLister
+}
+
+// NewAddFinalizerController returns a AddFinalizerController
+func NewAddFinalizerController(
```

Comment:
fwiw, I think that simply adding it to your resource is enough.  if a user changes it, they shot themselves in the foot, but I don't they expand their impact/control

Reply:
the role/rolebinding is currently applied by registration controller. Do you think it is better we should apply work's role/rolebindnig in work controller?

Reply:
The suggestion makes sense. Now the resources (role/rolebinding) are created by controller in registration repo. That means the finalizer will be added by controller in registration repo and removed by controller in work repo. This may not be a good idea. How about we still use AddFinalizerController to add finalizer to role/rolebinding until we move the creation of role/rolebinding from registration repo to work repo?

Reply:
> The suggestion makes sense. Now the resources (role/rolebinding) are created by controller in registration repo. That means the finalizer will be added by controller in registration repo and removed by controller in work repo. This may not be a good idea. How about we still use AddFinalizerController to add finalizer to role/rolebinding until we move the creation of role/rolebinding from registration repo to work repo

The matching logic here isn't significantly less fragile than simply adding the finalizer to the manifests in another repo.  Now, given usage, you may be better off putting this in the registration repo and simply listing the manifestworks until there are none.

Reply:
I'll create a PR on registration repo to make the change. But I'm still wondering if we have any use case which only involves registration repo. In that case, we may have trouble in deleting the role/rolebinding for work agent due to the finalizer.